File: CF1899G/CF1899G.go
package main

import (
	"bufio"
	. "fmt"
	"io"
	"os"
)

type node struct {
	lson int
	rson int
	cnt int
}

func solve(_r io.Reader, out io.Writer) {
	in := bufio.NewReader(_r)
	T := 1
	Fscan(in, &T)

	const N = 100010
	tr := make([]node, N * 4 + 17 * N)
	root := make([]int, N)
	dfn := make([]int, N)
	outn := make([]int, N)
	p := make([]int, N)
	idx, timestamp := 0, 0
	var insert func(int, int, int, int) int
	var query func(int, int, int, int, int, int) int
	insert = func(p, l, r, x int) int {
		idx ++
		q := idx
		if l == r {
			tr[q].cnt ++
			return q
		}
		tr[q] = tr[p]
		mid := (l + r) / 2
		if x <= mid {
			tr[q].lson = insert(tr[p].lson, l, mid, x)
		} else {
			tr[q].rson = insert(tr[p].rson, mid + 1, r, x)
		}
		tr[q].cnt = tr[tr[q].lson].cnt + tr[tr[q].rson].cnt
		return q
	}
	query = func(p, q, l, r, L, R int) int {
		if L <= l && r <= R {
			return tr[q].cnt - tr[p].cnt;
		}
		mid := (l + r) / 2
		res := 0
		if L <= mid {
			res += query(tr[p].lson, tr[q].lson, l, mid, L, R)
		}
		if R > mid {
			res += query(tr[p].rson, tr[q].rson, mid + 1, r, L, R)
		}
		return res
	}

	for ; T > 0; T-- {
		for i := 1; i < idx; i++ {
			tr[i].cnt, tr[i].lson, tr[i].rson = 0, 0, 0
		}
		idx, timestamp = 0, 0

		var n, q int
		Fscan(in, &n, &q)
		e := make([][]int, n + 1)
		for i := 1; i <= n; i++ {
			e[i] = make([]int, 0)
		}
		for i, u, v := 1, 0, 0; i < n; i++ {
			Fscan(in, &u, &v)
			e[u] = append(e[u], v)
			e[v] = append(e[v], u)
		}
		for i := 1; i <= n; i++ {
			Fscan(in, &p[i])
		}
		var dfs func(int, int)
		dfs = func(u, fa int) {
			timestamp ++
			dfn[u] = timestamp
			for _, v := range(e[u]) {
				if v != fa {
					dfs(v, u)
				}
			}
			outn[u] = timestamp
			// Fprintln(out, u, dfn[u], outn[u])
		}
		dfs(1, 0)
		for i := 1; i <= n; i++ {
			root[i] = insert(root[i - 1], 1, n, dfn[p[i]])
		}
		for i, l, r, x := 1, 0, 0, 0; i <= q; i++ {
			Fscan(in, &l, &r, &x)
			res := query(root[l - 1], root[r], 1, n, dfn[x], outn[x])
			if res > 0 {
				Fprintln(out, "YES")
			} else {
				Fprintln(out, "NO")
			}
		}
	}	
}

func main() { solve(os.Stdin, os.Stdout) }

File: cpp_code/CF1899G/CF1899G.cpp
#include <bits/stdc++.h>
#define x first
#define y second
#define sz(x) ((int)x.size())
using namespace std;

typedef long long ll;
const int inf = 0x3f3f3f3f;
const ll INF = 1e18;
const int N = 100010;

int idx;
int root[N], dfn[N], out[N], p[N];
struct node {
    int l, r, cnt;
} tr[N * 4 + N * 17];

int insert(int p, int l, int r, int x) {
    int q = ++ idx;
    tr[q] = tr[p];
    if (l == r) {
        tr[q].cnt ++;
        return q;
    }
    int mid = (l + r) / 2;
    if (x <= mid) {
        tr[q].l = insert(tr[p].l, l, mid, x);
    } else {
        tr[q].r = insert(tr[p].r, mid + 1, r, x);
    }
    tr[q].cnt = tr[tr[q].l].cnt + tr[tr[q].r].cnt;
    return q;
}

int query(int p, int l, int r, int L, int R) {
    if (L <= l && r <= R) {
        return tr[p].cnt;
    }

    int mid = (l + r) / 2;
    int res = 0;
    if (L <= mid) {
        res += query(tr[p].l, l, mid, L, R);
    }
    if (R > mid) {
        res += query(tr[p].r, mid + 1, r, L, R);
    }
    return res;
}

void solve() {
    int n, q, timestmp = 0;
    cin >> n >> q;
    vector<vector<int>> e(n + 1);
    for (int i = 1, u, v; i < n; i ++) {
        cin >> u >> v;
        e[u].push_back(v);
        e[v].push_back(u);
    }
    for (int i = 1; i <= n; i ++) {
        cin >> p[i];
    }
    auto dfs = [&](auto dfs, int u, int fa) -> void {
        dfn[u] = ++ timestmp;
        for (int v : e[u]) {
            if (v != fa) {
                dfs(dfs, v, u);
            }
        }
        out[u] = timestmp;
    };
    dfs(dfs, 1, 0);

    idx = 0;
    for (int i = 1; i <= n; i ++) {
        root[i] = insert(root[i - 1], 1, n, dfn[p[i]]);
    }
    while (q -- > 0) {
        int l, r, x;
        cin >> l >> r >> x;
        int cnt = query(root[r], 1, n, dfn[x], out[x]) - 
            query(root[l - 1], 1, n, dfn[x], out[x]);
        cout << (cnt > 0 ? "YES\n": "NO\n");
    }
    cout << '\n';
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr), cout.tie(nullptr);
    int t = 1;
    cin >> t;
    while (t--) solve();
    return 0;
}
